infra/token-transfer-server/src/lib/vesting.js

```diff
@@ -21,93 +21,123 @@ function momentizeGrant(grant) {
   const vestingEvents = [cliffVestAmount, ...remainingVestingAmounts]
+
+  // Add an rounding errors to the last vesting event
   const roundingError = BigNumber(grant.amount).minus(
     vestingEvents.reduce((a, b) => a.plus(b), BigNumber(0))
   )
   vestingEvents[vestingEvents.length - 1] = vestingEvents[
     vestingEvents.length - 1
   ].plus(roundingError)
 
-  return vestingEvents
-}
-
-/* Returns an array of vesting objects that include a datetime and status
- * associated with each vesting event.
- */
-function vestingSchedule(grantObj /* start = null, end = null */) {
-  const now = grantObj.now || moment()
-  const grant = momentizeGrant(grantObj)
-
-  // TODO implement filtering on start and end
-  /*
-  let bottomThreshold, topThreshold
-  if (start) {
-    bottomThreshold = start.diff(grant.cliff, grant.interval)
-  }
-  if (end) {
-    topThreshold = end.diff(grant.cliff, grant.interval)
-  }
-  */
-
-  return vestingEvents(grant).map((currentVestingEvent, index) => {
+  const events = vestingEvents.map((currentVestingEvent, index) => {
     const vestingEventDate = grant.cliff.clone()
     if (index > 0) {
-      vestingEventDate.add(index, grant.interval)
+      vestingEventDate.add(index, 'months')
     }
     return {
       amount: currentVestingEvent,
       date: vestingEventDate,
-      vested: vestingEventDate < now
+      vested: vestingEventDate <= now
     }
   })
+  return events
 }
 
-/* Returns the number of tokens vested by a grant.
- *
- */
-function vestedAmount(grantObj) {
-  const now = grantObj.now || moment()
+function investorVestingSchedule(grantObj) {
   const grant = momentizeGrant(grantObj)
+  const now = moment.utc()
+
+  const vestingSchedule = []
 
-  if (now < grant.cliff) {
-    return BigNumber(0)
-  } else if (now > grant.end) {
-    return BigNumber(grant.amount)
-  } else {
-    // Number of vesting events that have occurred since the cliff determines
-    // the index of the array for calculating events that have already vested
-    const threshold = now.diff(grant.cliff, grant.interval)
-    // Buld array of already vested amounts
-    const vested = vestingEvents(grant).slice(0, threshold + 1)
-    return vested.length ? Math.round(BigNumber.sum(...vested)) : BigNumber(0)
+  // Calculate initial vest percentage granted on grant start date
+  const initialVestPercentage = 6
+  // Time after which regular quarterly vesting begins
+  const quarterlyVestDelayMonths = 4
+  const firstQuarterlyVestDate = grant.start
+    .clone()
+    .add(quarterlyVestDelayMonths, 'months')
+  const quarterlyVestingPercentage = (100 - initialVestPercentage) / 8
+  const quarterlyVestingAmount = BigNumber(grant.amount)
+    .times(quarterlyVestingPercentage)
+    .div(100)
+
+  // Add initial vest
+  vestingSchedule.push({
+    amount: BigNumber(grant.amount)
+      .times(initialVestPercentage)
+      .div(100),
+    date: grant.start.clone(),
+    vested: grant.start <= now
+  })
+
+  // First quarterly vest
```

Comment:
nit: why doing this outside of the loop below?
could we have a single loop
```
const vestingDate = firstQuarterlyVestDate
for (let i=1; i<=8; i++) {
  vestingSchedule.push({
    amount: quarterlyVestingAmount,
    date: vestingDate.clone(),
    vested: vestingDate <= now
  })
  vestingDate.add(3, 'months')
}
```

Reply:
👍